### ISSUE TITLE
CVE-2021-25741: Symlink Exchange Can Allow Host Filesystem Access

### ISSUE THREAD

**Issue by miwithro** [Azure/AKS#2547]
A security issue was discovered in Kubernetes where a user may be able to create a container with subpath volume mounts to access files & directories outside of the volume, including on the host filesystem.
This issue has been rated High (CVSS:3.0/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:H), and assigned CVE-2021-25741.

#Affected Components and Configurations
This bug affects kubelet.

Environments where cluster administrators have restricted the ability to create hostPath mounts are the most seriously affected. Exploitation allows hostPath-like access without use of the hostPath feature, thus bypassing the restriction. 

In a default Kubernetes environment, exploitation could be used to obscure misuse of already-granted privileges.
Affected Versions
•	v1.22.0 - v1.22.1
•	v1.21.0 - v1.21.4
•	v1.20.0 - v1.20.10
•	<= v1.19.14
#Fixed Versions
This issue is fixed in the following versions:
•	v1.22.2
•	v1.21.5
•	v1.20.11
•	v1.19.15
#Mitigation
To mitigate this vulnerability without upgrading kubelet, you can disable the VolumeSubpath feature gate on kubelet and kube-apiserver, and remove any existing Pods making use of the feature.

You can also use admission control to prevent less-trusted users from running containers as root to reduce the impact of successful exploitation.
#Detection
If you find evidence that this vulnerability has been exploited, please contact security@kubernetes.io
Additional Details
See Kubernetes Issue #104980 for more details.

#AKS Information:

Ensure you upgrade to at least the [2021-08-26](https://github.com/Azure/AKS/releases/tag/2021-08-26) release that contains the patched kubelet  version. 

AKS
-- | --

**Comment by joberget** [comment]
@miwithro so no need to wait for example for version 1.22.2 or 1.21.5 in AKS? Kubelet is patched with version 5.4.0-1055-azure or newer of the kernel?

**Comment by miwithro** [comment]
No,  run a node image upgrade to the latest VHD and you are protected.

**Comment by tdihp** [comment]
Try `az aks list -ojson | jq '.[] |  .id as $id | .agentPoolProfiles[] | select(.osType!="Windows") |  {id:$id, pool:.name, nodeImageVersion:.nodeImageVersion, imageVersionDate:(.nodeImageVersion | match("[0-9.]{10}").string) } | select(.imageVersionDate < "2021.08.26")'`

**Comment by joberget** [comment]
@tdihp Thanks, but I get the following:

`jq: error (at <stdin>:1295): null (null) cannot be matched, as it is not a string`

**Comment by joberget** [comment]
Looks like this is related to:

`This cluster is not using VirtualMachineScaleSets. Node image upgrade only operation can only be applied on VirtualMachineScaleSets cluster.`

**Comment by miwithro** [comment]
@joberget Are you running VMAS?

**Comment by joberget** [comment]
Some clusters are running VMAS and some older are not using VMAS.

**Comment by miwithro** [comment]
@joberget is the node image upgrade failing on the VMSS nodes as well?

**Comment by joberget** [comment]
@miwithro how can we check if it is running the latest NodeImage on clusters without VMAS?

**Comment by miwithro** [comment]
@joberget you can run an az aks show -n -g 

 "nodeImageVersion": "AKSUbuntu-1804gen2containerd-2021.09.06"

or you can look in the portal

![image](https://user-images.githubusercontent.com/50749048/134517537-f2300836-0571-4c5f-b898-b01514703ace.png)

**Comment by joberget** [comment]
@miwithro  That value is empty here so it looks like it represents the same we see from az aks command where it says null.

**Comment by miwithro** [comment]
right that is expected in VMAS but not in VMSS, can you migrate to VMSS?

**Comment by joberget** [comment]
@miwithro Looks like this is a feature request? https://github.com/Azure/AKS/issues/2325

**Comment by tdihp** [comment]
@joberget thanks for the update.

I've modified the script to ignore vmas and sharing a separate script to filter out affected vmas nodes. Note that further efforts will be needed from user side to locate the affected AKS clusters from the VM shown.

Updated jq script to filter all VMSS based nodePools:

`az aks list -ojson | jq '.[] | .id as $id | .agentPoolProfiles[] | select(.osType!="Windows" and .type=="VirtualMachineScaleSets") | {id:$id, pool:.name, nodeImageVersion:.nodeImageVersion, imageVersionDate:(.nodeImageVersion | match("[0-9.]{10}").string) } | select(.imageVersionDate < "2021.08.26")'`

 Script to verify all VMSS with images provided by AKS "galleries/AKSUbuntu/images" gallery:

`az vm list -ojson | jq '.[] | select(.storageProfile.imageReference.id and (.storageProfile.imageReference.id | test("galleries/AKSUbuntu/images"))) | {id:.id, version: .storageProfile.imageReference.exactVersion} | select(.version<"2021.08.26")'`

**Comment by joberget** [comment]
@tdihp Thanks, works fine! But how do I upgrade nodes with VMAS?
```
# az aks upgrade --resource-group rg-cluster --name cluster --node-image-only            

Kubernetes may be unavailable during cluster upgrades.
 Are you sure you want to perform this operation? (y/N): y
This node image upgrade operation will run across every node pool in the clusterand might take a while, do you wish to continue? (y/N): y
This cluster is not using VirtualMachineScaleSets. Node image upgrade only operation can only be applied on VirtualMachineScaleSets cluster.
```

**Comment by tdihp** [comment]
@joberget for VMAS based clusters, node-image-only doesn't work. I'd suggest to use below alternatives:

1. If possible to upgrade (i.e. the current node k8s version is not latest), trigger a k8s version upgrade. New VMs will all have latest nodeImages, and old VMs will be retired in the process.
2. If not possible to upgrade:
  1). Scale up new VMs which will have latest images.
  2). Drain the vulnerable nodes with kubectl.
  3). After drain successful, delete the vulnerable VMs in the node resource group for the vulnerable nodes
  4). Then issue a separate scale to match resource node count with the actual node count.

Note that option 2 operates VMs directly which is not a supported or recommended action for AKS resources. It is only suggested as ad-hoc mitigation if upgrade is not possible.

**Comment by joberget** [comment]
@tdihp Thanks, is there any way to migrate from VMAS to VMSS or do I need to setup a new cluster?

**Comment by tdihp** [comment]
@joberget AKS currently only offers the choice of VMAS/VMSS on creation unfortunately :(.
